contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,652 @@
+  // so in the worse case where the consuming contract has to read all of them
+  // from storage, it only has to read 2 words.
+  function requestRandomWords(
+    bytes32 keyHash,  // Corresponds to a particular offchain job which uses that key for the proofs
+    uint64  subId,
+    uint16  requestConfirmations,
+    uint32  callbackGasLimit,
+    uint32  numWords,  // Desired number of random words
+    uint32  consumerId // Index into consumers to avoid SLOADing all the consumers
+  )
+    external
+    returns (
+      uint256 requestId
+    )
+  {
+    // Input validation using the subscription storage.
+    if (s_subscriptions[subId].owner == address(0)) {
+      revert InvalidSubscription();
+    }
+    // We use this consumer index to ensure that the cost remains constant no matter
+    // the number of consumers.
+    // Its important to ensure that the consumer is in fact who they say they
+    // are, otherwise they could use someone else's subscription balance.
+    if (s_subscriptions[subId].consumers[consumerId] != msg.sender) {
+      revert InvalidConsumer(msg.sender);
+    }
+    // Input validation using the config storage word.
+    if (requestConfirmations < s_config.minimumRequestBlockConfirmations || requestConfirmations > MAX_REQUEST_CONFIRMATIONS) {
+      revert InvalidRequestBlockConfs(requestConfirmations, s_config.minimumRequestBlockConfirmations, MAX_REQUEST_CONFIRMATIONS);
+    }
+    if (s_subscriptions[subId].balance < s_config.minimumSubscriptionBalance) {
+      revert InsufficientBalance();
+    }
+    if (callbackGasLimit > s_config.maxGasLimit) {
+      revert GasLimitTooBig(callbackGasLimit, s_config.maxGasLimit);
+    }
+    // We could additionally check s_serviceAgreements[keyHash] != address(0)
+    // but that would require reading another word of storage. To save gas
+    // we leave that out.
+    uint256 nonce = s_nonces[keyHash][msg.sender] + 1;
+    uint256 preSeedAndRequestId = uint256(keccak256(abi.encode(keyHash, msg.sender, nonce)));
+
+    s_callbacks[preSeedAndRequestId] = keccak256(abi.encodePacked(
+        preSeedAndRequestId,
+        block.number,
+        subId,
+        callbackGasLimit,
+        numWords,
+        msg.sender));
+    emit RandomWordsRequested(keyHash, preSeedAndRequestId, subId, requestConfirmations, callbackGasLimit, numWords, msg.sender);
+    s_nonces[keyHash][msg.sender] = nonce;
+
+    return preSeedAndRequestId;
+  }
+
+  function getCallbackHash(
+    uint256 requestId
+  )
+    external
+    view
+    returns (
+      bytes32
+    )
+  {
+    return s_callbacks[requestId];
+  }
+
+  /**
+   * @dev calls target address with exactly gasAmount gas and data as calldata
+   * or reverts if at least gasAmount gas is not available
+   */
+  function callWithExactGas(
+    uint256 gasAmount,
+    address target,
+    bytes memory data
+  )
+    private
+    returns (
+      bool success
+    )
+  {
+    assembly{
```

Comment:
A lot going on here; where is this from and have we got tests for it?

Reply:
yeah this is lifted from the keeper contracts directly

Reply:
Could we use a test helper contract to import those tests in this repo too?

Reply:
@RyanRHall are there plans to bring the keeper contracts over here soon?